kinesis/kinesis.go

```diff
@@ -210,6 +213,10 @@ func (outputPlugin *OutputPlugin) Flush(records *[]*kinesis.PutRecordsRequestEnt
 				logrus.Errorf("[kinesis %d] %v\n", outputPlugin.PluginID, err)
 			}
 			if retCode != fluentbit.FLB_OK {
+				unsent := (*records)[i:]
+				// requestBuf will contain records sendCurrentBatch failed to send,
+				// combine those with the records yet to be sent/batched
+				*records = append(requestBuf, unsent...)
```

Comment:
I'm not sure about this logic- can you help make sure I understand- there's a case where `processAPIResponse` will already append the failed records to the record list (if the Kinesis API returns partial success). In that specific case, doesn't this still lead to duplicates?

Reply:
I noticed when we were receiving backpressure from kinesis larges blocks of records were resent I traced it to this logic.  This logic prevents the log duplication:

Below is what would happen before these changes.

```

records = [ 'a', 'b', 'c', 'd']
flushWithRetries( records )  // records = [ 'a', 'b', 'c', 'd']
|-> pluginConcurrentFlush( records )  // records = [ 'a', 'b', 'c', 'd']
      |-> Flush(records)  // records = [ 'a', 'b', 'c', 'd']
            |->  sendCurrentBatch(record) // records = [ 'a', 'b', 'c', 'd']
                  // Assume maximumPutRecordBatchSize is 2 for illustration purposes
                 requestBuf = [ 'a', 'b' ] // Flush builds a records list named requestBuf with size maximumPutRecordBatchSize
                  // sendCurrentBatch fails to send record 'b'
                   |-> processAPIResponse( records/requestBuf )  // records/requestBuf = [ 'a', 'b' ]
                          // processAPIResponse updates records to reflect that 'a' was send, but 'b' wasn't
                          records/requestBuf = [ 'b' ]
                   <-|  processAPIResponse
                   // return to sendCurrentBatch  records/requestBuf = [ 'b' ]
            <-| sendCurrentBatch
             // return to Flush  records/requestBuf = [ 'b' ]
         <-| Flush 
          // return to pluginConcurrentFlush without update the records value passed 
   <-| pluginConcurrentFlush
    // Since requestBuf is in the scope of Flush pluginConcurrentFlush still has a slice of  records = [ 'a', 'b', 'c', 'd']
    // On the next retry the entire block of records is retried not just [ 'b', 'c', 'd']
    ...
```

Reply:
Following these changes this is the management of the records list:

```
records = [ 'a', 'b', 'c', 'd']
flushWithRetries( records )  // records = [ 'a', 'b', 'c', 'd']
|-> pluginConcurrentFlush( records )  // records = [ 'a', 'b', 'c', 'd']
      |-> Flush(records)  // records = [ 'a', 'b', 'c', 'd']
            |->  sendCurrentBatch(record) // records = [ 'a', 'b', 'c', 'd']
                  // Assume maximumPutRecordBatchSize is 2 for illustration purposes
                 requestBuf = [ 'a', 'b' ] // Flush builds a records list named requestBuf with size maximumPutRecordBatchSize
                  // sendCurrentBatch fails to send record 'b'
                   |-> processAPIResponse( records/requestBuf )  // records/requestBuf = [ 'a', 'b' ]
                          // processAPIResponse updates records to reflect that 'a' was send, but 'b' wasn't
                          records/requestBuf = [ 'b' ]
                   <-|  processAPIResponse
                   // return to sendCurrentBatch  records/requestBuf = [ 'b' ]
            <-| sendCurrentBatch
             // return to Flush  records/requestBuf = [ 'b' ]
         <-| Flush 
          // return to pluginConcurrentFlush by combining the records in requestBuf with the unsent records
          // Set records to [ 'b', 'c', 'd']
   <-| pluginConcurrentFlush
    // Now  records = [ 'b', 'c', 'd']
    // On the next retry only the records that failed, or were not tried are sent. records = [ 'b', 'c', 'd']
    ...
```